sql/core/src/main/scala/org/apache/spark/sql/Observation.scala

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+
+import org.apache.spark.sql.execution.QueryExecution
+import org.apache.spark.sql.util.QueryExecutionListener
+
+/**
+ * Helper class to simplify usage of [[Dataset.observe(String, Column, Column*)]]:
+ *
+ * {{{
+ *   // Observe row count (rows) and highest id (maxid) in the Dataset while writing it
+ *   val observation = Observation("my_metrics")
+ *   val observed_ds = ds.observe(observation, count(lit(1)).as("rows"), max($"id").as("maxid"))
+ *   observed_ds.write.parquet("ds.parquet")
+ *   val metrics = observation.get
+ * }}}
+ *
+ * This collects the metrics while the first action is executed on the obseerved dataset. Subsequent
+ * actions do not modify the metrics returned by [[org.apache.spark.sql.Observation.get]]. Retrieval
+ * of the metric via [[org.apache.spark.sql.Observation.get]] blocks until the first action has
+ * finished and metrics become available. You can add a timeout to that blocking via
+ * [[org.apache.spark.sql.Observation.waitCompleted]]:
+ *
+ * {{{
+ *   if (observation.waitCompleted(100, TimeUnit.MILLISECONDS)) {
+ *     observation.get
+ *   }
+ * }}}
+ *
+ * This class does not support streaming datasets.
+ *
+ * @param name name of the metric
+ * @since 3.2.0
+ */
+class Observation(name: String) {
```

Comment:
This is in essence an asynchronous API. You never know when the computation actually finishes, and in most cases you are only interested in it when it produces a result. So why are we introducing concepts like wait & get? To me it would be more logical to make this return a Future. In that case the user can compose a further computation using that & if they need the blocking API they can just use Future.get/await.

Reply:
That's a good point! We can simply leverage the standard `Future` API to represent this async result.

Reply:
This `Observation` is trying hiding multi-threading from the user for use cases where they simply want a metric after they have called an action on a Dataset. Adding async notion to the API does not contribute to that goal.

The `waitCompleted` is an edge case, usually you simply call `Observation.get`. If the user wanted a Future, they could wrap the `Observation.get` with one. What you are proposing would mean that everyone else who does not want a `Future` would get a `Future` and has to call `.value` to get an `Option[Try[Row]]` and deal with this result where `Observation.get` now returns the `Row`. I do not see where the Future simplifies the API.

Maybe you can sketch this out a bit more on how it would look like for the user of `Observation` so the benefits become clearer.

Reply:
We can still keep the `get` API, but returning `Future` should be a more standard way for users to do async than `waitCompleted`.

Reply:
We can remove the public `waitCompleted` and only provide `get`. If they want a `Future`, they can

    val f = Future { observation.get }

and they want a timeout, they can do:

    Await.result(f, 2 seconds)

Reply:
SGTM

Reply:
done